## System
- Hot Pepper Gourmet(hpg): 이용자들이 식당을 찾고 온라인 예약할 수 있는 시스템
- AirREGI / Restaurant Board(air): 예약 관리 및 계산 시스템

## air_reserve.csv
- air_store_id: 식당의 ID(air)
- visit_datetime: 예약 시간(방문 예정 시간)
- reserve_datetime: 예약 등록을 한 시간
- reserve_visitors: 예정 방문자 수

## hpg_reserve.csv
- hpg_store_id: 식당의 ID(hpg)
- visit_datetime: 예약 시간(방문 예정 시간)
- reserve_datetime: 예약 등록을 한 시간
- reserve_visitors: 예정 방문자 수

## air_store_info.csv
식당에서 air 시스템에 등록한 정보
- air_store_id: 식당 ID(air)
- air_genre_name: 식당의 종류(ex.이탈리안, 이자카야) 
- air_area_name: 지역
- latitude: 경도 (해당 지역의 위도, 경도)
- longitude: 위도

## hpg_store_info.csv
식당에서 hpg 시스템에 등록한 정보
- air_store_id: 식당 ID(hpg)
- air_genre_name: 식당의 종류(ex.이탈리안, 이자카야) 
- air_area_name: 지역
- latitude: 경도 (해당 지역의 위도, 경도)
- longitude: 위도

## store_id_relation.csv
두 시스템에서의 ID가 매칭되어 있다.
- hpg_store_id
- air_store_id

## air_visit_data.csv
방문 기록이 담겨있다.
- air_store_id: 식당의 ID(air)
- visit_date - 방문 날짜
- visitors - 방문자의 수

## date_info.csv
- calendar_date: 날짜(년, 월, 일)
- day_of_week: 요일
- holiday_flg - 일본 기준 휴일 여부(1,0)

## sample_submission.csv
제출 양식
- id: air 기준 식당 ID와 방문 날짜가 결합된 형태
- visiotors: id에 해당하는 날짜, 식당에 방문할 손님의 수 예측


# 데이터 불러오기 및 확인

In [68]:
import pandas as pd

air_reserve = pd.read_csv("air_reserve.csv")
hpg_reserve = pd.read_csv("hpg_reserve.csv")
air_store_info = pd.read_csv("air_store_info.csv")
hpg_store_info = pd.read_csv("hpg_store_info.csv")
store_id_relation = pd.read_csv("store_id_relation.csv")
air_visit_data = pd.read_csv("air_visit_data.csv")
date_info = pd.read_csv("date_info.csv")
stores_to_pred = pd.read_csv("sample_submission.csv")


## air_store_info

In [19]:
air_store_info.shape

(829, 5)

In [18]:
air_store_info.head()

air_store_id  air_genre_name                 air_area_name  \
0  air_0f0cdeee6c9bf3d7  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
1  air_7cc17a324ae5c7dc  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
2  air_fee8dcf4d619598e  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
3  air_a17f0778617c76e2  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
4  air_83db5aff8f50478e  Italian/French  Tōkyō-to Minato-ku Shibakōen   

    latitude   longitude  
0  34.695124  135.197853  
1  34.695124  135.197853  
2  34.695124  135.197853  
3  34.695124  135.197853  
4  35.658068  139.751599

In [28]:
# 중복값 확인
air_store_info['air_store_id'].duplicated().sum()

0

## hpg_store_info

In [20]:
hpg_store_info.shape

(4690, 5)

In [9]:
hpg_store_info.head()

hpg_store_id  hpg_genre_name                  hpg_area_name  \
0  hpg_6622b62385aec8bf  Japanese style  Tōkyō-to Setagaya-ku Taishidō   
1  hpg_e9e068dd49c5fa00  Japanese style  Tōkyō-to Setagaya-ku Taishidō   
2  hpg_2976f7acb4b3a3bc  Japanese style  Tōkyō-to Setagaya-ku Taishidō   
3  hpg_e51a522e098f024c  Japanese style  Tōkyō-to Setagaya-ku Taishidō   
4  hpg_e3d0e1519894f275  Japanese style  Tōkyō-to Setagaya-ku Taishidō   

    latitude   longitude  
0  35.643675  139.668221  
1  35.643675  139.668221  
2  35.643675  139.668221  
3  35.643675  139.668221  
4  35.643675  139.668221

In [30]:
# 중복값 확인
hpg_store_info['hpg_store_id'].duplicated().sum()

0

air 시스템에는 829개, hpg 시스템에는 4690개의 식당이 등록되어 있습니다.

## air_reserve

In [31]:
air_reserve.shape

(92378, 4)

In [52]:
air_reserve.head()

air_store_id       visit_datetime     reserve_datetime  \
0  air_877f79706adbfb06  2016-01-01 19:00:00  2016-01-01 16:00:00   
1  air_db4b38ebe7a7ceff  2016-01-01 19:00:00  2016-01-01 19:00:00   
2  air_db4b38ebe7a7ceff  2016-01-01 19:00:00  2016-01-01 19:00:00   
3  air_877f79706adbfb06  2016-01-01 20:00:00  2016-01-01 16:00:00   
4  air_db80363d35f10926  2016-01-01 20:00:00  2016-01-01 01:00:00   

   reserve_visitors  
0                 1  
1                 3  
2                 6  
3                 2  
4                 5

In [57]:
display(air_reserve.sample(10))

air_store_id       visit_datetime     reserve_datetime  \
71890  air_0a74a5408a0b8642  2017-03-10 18:00:00  2017-03-08 15:00:00   
7672   air_e55abd740f93ecc4  2016-03-22 17:00:00  2016-03-22 00:00:00   
72316  air_bc991c51d6613745  2017-03-10 20:00:00  2017-03-08 17:00:00   
13297  air_f911308e19d64236  2016-05-15 13:00:00  2016-05-11 13:00:00   
31548  air_1408dd53f31a8a65  2016-12-07 21:00:00  2016-12-06 19:00:00   
34353  air_df554c4527a1cfe6  2016-12-13 20:00:00  2016-11-30 21:00:00   
54117  air_55c3627912b9c849  2017-01-27 18:00:00  2017-01-07 00:00:00   
50327  air_55c3627912b9c849  2017-01-17 19:00:00  2017-01-17 19:00:00   
33714  air_a083834e7ffe187e  2016-12-11 20:00:00  2016-12-11 18:00:00   
41738  air_4092cfbd95a3ac1b  2016-12-25 20:00:00  2016-12-25 15:00:00   

       reserve_visitors  
71890                 2  
7672                  3  
72316                16  
13297                 2  
31548                 3  
34353                 2  
54117                 2  
50327                 4  
33714                 2  
41738                 2

In [58]:
air_reserve.duplicated().sum()

2770

In [41]:
air_reserve[air_reserve.duplicated()]

air_store_id       visit_datetime     reserve_datetime  \
48     air_789466e488705c93  2016-01-04 17:00:00  2016-01-04 17:00:00   
92     air_e55abd740f93ecc4  2016-01-05 20:00:00  2016-01-05 18:00:00   
122    air_632ba66e1f75aa28  2016-01-06 19:00:00  2016-01-06 15:00:00   
163    air_de692863bb2dd758  2016-01-07 18:00:00  2016-01-07 18:00:00   
196    air_03963426c9312048  2016-01-08 17:00:00  2016-01-06 15:00:00   
...                     ...                  ...                  ...   
91548  air_db4b38ebe7a7ceff  2017-04-29 17:00:00  2017-04-06 22:00:00   
91611  air_640cf4835f0d9ba3  2017-04-29 19:00:00  2017-03-30 17:00:00   
91874  air_900d755ebd2f7bbd  2017-05-04 12:00:00  2017-04-19 16:00:00   
92157  air_900d755ebd2f7bbd  2017-05-12 18:00:00  2017-04-19 16:00:00   
92266  air_8e492076a1179383  2017-05-17 18:00:00  2017-04-18 20:00:00   

       reserve_visitors  
48                    2  
92                    2  
122                   2  
163                   2  
196                   4  
...                 ...  
91548                 2  
91611                 2  
91874                 3  
92157                 6  
92266                 2  

[2770 rows x 4 columns]

In [55]:
air_reserve[(air_reserve['air_store_id'] == 'air_789466e488705c93') & (air_reserve['visit_datetime'] == '2016-01-04 17:00:00') ]

air_store_id       visit_datetime     reserve_datetime  \
47  air_789466e488705c93  2016-01-04 17:00:00  2016-01-04 17:00:00   
48  air_789466e488705c93  2016-01-04 17:00:00  2016-01-04 17:00:00   

    reserve_visitors  
47                 2  
48                 2

In [56]:
air_reserve[(air_reserve['air_store_id'] == 'air_e55abd740f93ecc4') & (air_reserve['visit_datetime'] == '2016-01-05 20:00:00') ]

air_store_id       visit_datetime     reserve_datetime  \
91  air_e55abd740f93ecc4  2016-01-05 20:00:00  2016-01-05 18:00:00   
92  air_e55abd740f93ecc4  2016-01-05 20:00:00  2016-01-05 18:00:00   

    reserve_visitors  
91                 2  
92                 2

## hpg_reserve

In [43]:
hpg_reserve.shape

(2000320, 4)

In [7]:
hpg_reserve.head()

hpg_store_id       visit_datetime     reserve_datetime  \
0  hpg_c63f6f42e088e50f  2016-01-01 11:00:00  2016-01-01 09:00:00   
1  hpg_dac72789163a3f47  2016-01-01 13:00:00  2016-01-01 06:00:00   
2  hpg_c8e24dcf51ca1eb5  2016-01-01 16:00:00  2016-01-01 14:00:00   
3  hpg_24bb207e5fd49d4a  2016-01-01 17:00:00  2016-01-01 11:00:00   
4  hpg_25291c542ebb3bc2  2016-01-01 17:00:00  2016-01-01 03:00:00   

   reserve_visitors  
0                 1  
1                 3  
2                 2  
3                 5  
4                13

In [45]:
hpg_reserve.duplicated().sum()

16279

## store_id_relation

In [47]:
store_id_relation.shape

(150, 2)

In [10]:
store_id_relation.head()

air_store_id          hpg_store_id
0  air_63b13c56b7201bd9  hpg_4bc649e72e2a239a
1  air_a24bf50c3e90d583  hpg_c34b496d0305a809
2  air_c7f78b4f3cba33ff  hpg_cd8ae0d9bbd58ff9
3  air_947eb2cae4f3e8f2  hpg_de24ea49dc25d6b8
4  air_965b2e0cf4119003  hpg_653238a84804d8e7

In [46]:
store_id_relation.duplicated().sum()

0

## air_visit_data

In [48]:
air_visit_data.shape

(252108, 3)

In [13]:
air_visit_data.head()

air_store_id  visit_date  visitors
0  air_ba937bf13d40fb24  2016-01-13        25
1  air_ba937bf13d40fb24  2016-01-14        32
2  air_ba937bf13d40fb24  2016-01-15        29
3  air_ba937bf13d40fb24  2016-01-16        22
4  air_ba937bf13d40fb24  2016-01-18         6

In [49]:
air_visit_data.duplicated().sum()

0

## data_info

In [12]:
date_info.head()

calendar_date day_of_week  holiday_flg
0    2016-01-01      Friday            1
1    2016-01-02    Saturday            1
2    2016-01-03      Sunday            1
3    2016-01-04      Monday            0
4    2016-01-05     Tuesday            0

## stores to pred

In [70]:
stores_to_pred.head()

id  visitors
0  air_00a91d42b08b08d9_2017-04-23         0
1  air_00a91d42b08b08d9_2017-04-24         0
2  air_00a91d42b08b08d9_2017-04-25         0
3  air_00a91d42b08b08d9_2017-04-26         0
4  air_00a91d42b08b08d9_2017-04-27         0

In [71]:
id_split = stores_to_pred['id'].str.split('_', expand=True)
id_split.columns = ['source', 'store_id', 'visit_date']
stores_to_pred = pd.concat([stores_to_pred, id_split], axis=1)
stores_to_pred['air_store_id']=stores_to_pred['source'].str.cat(stores_to_pred['store_id'], sep='_')
stores_to_pred = stores_to_pred[['id', 'air_store_id', 'visit_date', 'visitors']]
stores_to_pred['visit_date'] = pd.to_datetime(stores_to_pred['visit_date'])

stores_to_pred.head()

id          air_store_id visit_date  visitors
0  air_00a91d42b08b08d9_2017-04-23  air_00a91d42b08b08d9 2017-04-23         0
1  air_00a91d42b08b08d9_2017-04-24  air_00a91d42b08b08d9 2017-04-24         0
2  air_00a91d42b08b08d9_2017-04-25  air_00a91d42b08b08d9 2017-04-25         0
3  air_00a91d42b08b08d9_2017-04-26  air_00a91d42b08b08d9 2017-04-26         0
4  air_00a91d42b08b08d9_2017-04-27  air_00a91d42b08b08d9 2017-04-27         0

In [72]:
stores_to_pred.air_store_id.nunique()

821

## Merge Store

In [74]:
stores_to_pred['air_store_id'].isin(air_store_info['air_store_id']).all()

True

예측하고자 하는 식당의 ID가 전부 air_store_id에 속해있으므로 air_store_id를 기준으로 병합을 해주어도 괜찮습니다.

In [60]:
stores = pd.merge(air_store_info, store_id_relation, how='left', on='air_store_id')
stores.shape

(829, 6)

In [64]:
stores.head()

air_store_id  air_genre_name                 air_area_name  \
0  air_0f0cdeee6c9bf3d7  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
1  air_7cc17a324ae5c7dc  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
2  air_fee8dcf4d619598e  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
3  air_a17f0778617c76e2  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
4  air_83db5aff8f50478e  Italian/French  Tōkyō-to Minato-ku Shibakōen   

    latitude   longitude          hpg_store_id  
0  34.695124  135.197853                   NaN  
1  34.695124  135.197853  hpg_9b38b9e13da6da27  
2  34.695124  135.197853                   NaN  
3  34.695124  135.197853                   NaN  
4  35.658068  139.751599                   NaN

In [66]:
stores = pd.merge(stores, hpg_store_info,how='left', on='hpg_store_id', suffixes=['_air', '_hpg'])

In [67]:
stores.head()

air_store_id  air_genre_name                 air_area_name  \
0  air_0f0cdeee6c9bf3d7  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
1  air_7cc17a324ae5c7dc  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
2  air_fee8dcf4d619598e  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
3  air_a17f0778617c76e2  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
4  air_83db5aff8f50478e  Italian/French  Tōkyō-to Minato-ku Shibakōen   

   latitude_air  longitude_air          hpg_store_id hpg_genre_name  \
0     34.695124     135.197853                   NaN            NaN   
1     34.695124     135.197853  hpg_9b38b9e13da6da27            NaN   
2     34.695124     135.197853                   NaN            NaN   
3     34.695124     135.197853                   NaN            NaN   
4     35.658068     139.751599                   NaN            NaN   

  hpg_area_name  latitude_hpg  longitude_hpg  
0           NaN           NaN            NaN  
1           NaN           NaN            NaN  
2           NaN           NaN            NaN  
3           NaN           NaN            NaN  
4           NaN           NaN            NaN